pkg/controller/machine-set-boot-image/machine_set_boot_image_controller.go

```diff
@@ -0,0 +1,685 @@
+	if !releaseVersionFound {
+		klog.Infof("failed to find mco release version in %s configmap, sync will exit to wait for the MCO upgrade to complete", ctrlcommon.BootImagesConfigMapName)
+		return nil
+	}
+	if releaseVersionFromCM != operatorversion.ReleaseVersion {
+		klog.Infof("mismatch between MCO release version stored in configmap and current MCO release version; sync will exit to wait for the MCO upgrade to complete")
+		return nil
+	}
+
+	// TODO: Also check against the release version stored in the configmap under releaseVersion. This is currently broken as the version
+	// stored is "0.0.1-snapshot" and does not reflect the correct value. Tracked in this bug https://issues.redhat.com/browse/OCPBUGS-19824
+	// The current hash and version check should be enough to skate by for now, but fixing this would be additional safety - djoshy
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// Fetch the machineset
+	machineSet, err := ctrl.machineSetLister.MachineSets(namespace).Get(name)
+	if machineSet == nil || apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to fetch machineset object during machineset sync: %w", err)
+	}
+
+	// Fetch the architecture type of this machineset
+	arch, err := ctrl.getArchFromMachineSet(machineSet)
+	if err != nil {
+		return fmt.Errorf("failed to fetch architecture type of machineset %s, err: %w", machineSet.Name, err)
+	}
+
+	// Fetch the infra object to determine the platform type
+	infra, err := ctrl.infraLister.Get("cluster")
+	if err != nil {
+		return fmt.Errorf("failed to fetch infra object during machineset sync: %w", err)
+	}
+
+	// Check if the this MachineSet requires an update
+	patchRequired, newMachineSet, err := checkMachineSet(infra, machineSet, configMap, arch)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile machineset %s, err: %w", machineSet.Name, err)
+	}
+
+	// Patch the machineset if required
+	if patchRequired {
+		klog.Infof("Patching machineset %s", machineSet.Name)
+		ctrl.patchMachineSet(machineSet, newMachineSet)
+	} else {
+		klog.Infof("No patching required for machineset %s", machineSet.Name)
+	}
+	return nil
+}
+
+// Returns architecture type for a given machine set
+func (ctrl *Controller) getArchFromMachineSet(machineset *machinev1beta1.MachineSet) (arch string, err error) {
+	machineSelector, err := metav1.LabelSelectorAsSelector(&machineset.Spec.Selector)
+
+	if err != nil {
+		return " ", fmt.Errorf("could not convert MachineSet label selector to selector, error: %w", err)
+	}
+
+	machines, err := ctrl.machineLister.List(machineSelector)
+	if err != nil || len(machines) == 0 {
+		return " ", fmt.Errorf("could not find any machines linked to machineset, error: %w", err)
+	}
+
+	// Any machine from this slice will be of the same architecture, so grab the first one
+	// Cycle through nodes, compare to annotations to find node with matching machine
+
+	nodes, err := ctrl.nodeLister.List(labels.Everything())
```

Comment:
We already chatted about this, and I think our plan is to read it off the label first, then either the nodes here or from the control plane to match autoscaler, but I wonder if we can skip the node check altogether and just follow the same steps outlined there?

Reply:
We totally could, and I was debating that as I was reworking this. The labels were present in my regular 3+3 cluster today, so I think it may have been a newer change that I hadn't caught on to yet.

Reply:
Opened a PR which resolves this: https://github.com/openshift/machine-config-operator/pull/4088
This should also fix https://issues.redhat.com/browse/OCPBUGS-25725